### ISSUE TITLE
Transactions used where they dont need to be

### ISSUE THREAD

**Issue by lridgard** [sqlitebrowser/sqlitebrowser#254]
EXPLAIN QUERY PLAN SELECT BlahField FROM BladDb WHERE foo=0;

begins a transaction when one is not required.

When using the DB browser to execute certain queries it enables the "write changes" and "revert changes" options when this is not required. This means DB Browser creates a transaction that is awaiting a COMMIT or ROLLBACK but no such thing is required in this case since the explain verb does not write anything. In journal DELETE mode an empty journal is created.

SELECT BlahField FROM BladDb WHERE foo=0;

works as expected and no transaction is creating awaiting a COMMIT or ROLLBACK.

**Comment by justinclift** [comment]
Yeah, I'm not sure how to solve this.  A dodgy workaround :wink: is to have "end transaction;" before your statement.  This closes our implicit transaction, then executes the statement after.

[Example here](https://github.com/sqlitebrowser/sqlitebrowser/issues/234#issuecomment-84955263), if that helps.

**Comment by lridgard** [comment]
Thanks for the quick response.
Adding "END TRANSACTION" can have catastrophic consequences if my main app is running at the same time which sadly is how I found this problem in the first place.

But what are your thoughts on one of these fixes:

Option 1: It appears the sqlitebrowser understands that there are times when a transaction is not required.
SELECT BlahField FROM BladDb WHERE foo=0;
executes without enabling the "write changes" and "revert changes" buttons. So since sqlitebrowser knows not to start a transaction in that case why not add statements like "EXPLAIN" to the list of cases where a transaction is not created similar to how "SELECT" is handled? In fact instead of having a hard-coded list, as I am suggesting, why not call
int sqlite3_stmt_readonly(sqlite3_stmt *pStmt);
on the prepared statement to determine if changes will be made by that statement. Then only start a transaction in the cases where changes will be made.

Option 2: Since Transactions cant be nested perhaps DB Browser should consider using (SAVEPOINT, RELEASE, ROLLBACK TO) instead of (BEGIN, COMMIT, ROLLBACK). This might work better in the cases where I am using sqlitebrowser to look up something in the DB while my main app is running. Note to self: I should perhaps make this same switch in my app too.

Option 3: Let the user control the creation of transactions. Add an option to disable the current transaction approach built into sqlitebrowser thus forcing us users, who choose that option, to manually create transactions via the BEGIN or SAVEPOINT in those cases where we feel it is required. And if we don't then we get the default SQLite automatic transactions on every statement.

Forgive me is any of this is already in place or has already been discounted as crazy. I have not looked through the code.

**Comment by rp-** [comment]
> Adding "END TRANSACTION" can have catastrophic consequences if my main app is running at the same time which sadly is how I found this problem in the first place.

How is that possible? different database connections cannot interfere each other?

But option1 with sqlite3_stmt_readonly sounds reasonable.

**Comment by justinclift** [comment]
@lridgard Any interest in submitting a pull request with the change? :smile:

**Comment by lridgard** [comment]
@rp- I have always thought that SQLite handled multiple processes accessing the DB perfectly fine. But after fooling around with the explain query in SQLiteBrowser I eventually wound up at a place where my main app stopped running complaining that it could not prepare a SQL statement. Closing the DB in SqliteBrowser resolved the issue and allowed my app to continue. Maybe this is something unrelated to the unrequired transactions topic so I should do some further investigation.

@justinclift Yes of course I would be willing to contribute some code. I am not familiar with the ins and outs of contributing code on GitHub but I am sure I can figure it out. If the desire is for Option 1 I can certainly put something together. BTW: in another thread I saw you were looking for a Visual Studio person to help with 32 to 64-bit code. I am willing to help there too. I currently build SQLite3 from the source into a 64-bit DLL that gets used by my 64-bit app. I am a windows coder.

**Comment by justinclift** [comment]
Oh wow.  Yeah, all of that sounds good. :grin:

With the situation you mention of the main app complaining... I suspect it might be something to do with the way SQLite locks tables to do writes.  It's been ages since I looked at the fine details though, so that's a rough guess thing. :wink:

**Comment by justinclift** [comment]
Hmmmm... a thought just occurred...

How about in the "Execute SQL" window, we have a line with a checkbox somewhere like:

&nbsp; &nbsp; [✓] Run inside transaction

We leave it on by default, but that would let people turn it off as needed.

?

**Comment by lridgard** [comment]
@justinclift  Adding such a checkbox would be fantastic. Leaving it on by default would preserve the current behavior and when people needed it to be off they could turn it off. Perfect.

**Comment by rp-** [comment]
@lridgard I already did the msvc stuff in https://github.com/sqlitebrowser/sqlitebrowser/commit/8cbc13a106a680e7e1ce97e2e49bb85eb80e9d8f so if you have the dependencies and the paths correctly set in cmake it should work.
